File: AirshipKit/AirshipKit/UAPreferenceDataStore.m
/*
 Copyright 2009-2016 Urban Airship Inc. All rights reserved.

 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions are met:

 1. Redistributions of source code must retain the above copyright notice, this
 list of conditions and the following disclaimer.

 2. Redistributions in binary form must reproduce the above copyright notice,
 this list of conditions and the following disclaimer in the documentation
 and/or other materials provided with the distribution.

 THIS SOFTWARE IS PROVIDED BY THE URBAN AIRSHIP INC ``AS IS'' AND ANY EXPRESS OR
 IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
 EVENT SHALL URBAN AIRSHIP INC OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
 INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
 OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#import "UAPreferenceDataStore+Internal.h"

@interface UAPreferenceDataStore()
@property (nonatomic, strong) NSUserDefaults *defaults;
@property (nonatomic, copy) NSString *keyPrefix;
@end


@implementation UAPreferenceDataStore

+ (instancetype)preferenceDataStoreWithKeyPrefix:(NSString *)keyPrefix {
    UAPreferenceDataStore *dataStore = [[UAPreferenceDataStore alloc] init];
    dataStore.defaults = [NSUserDefaults standardUserDefaults];
    dataStore.keyPrefix = keyPrefix;
    return dataStore;
}

- (NSString *)prefixKey:(NSString *)key {
    return [self.keyPrefix stringByAppendingString:key];
}

- (id)valueForKey:(NSString *)key {
    return [self.defaults valueForKey:[self prefixKey:key]];
}

- (void)setValue:(id)value forKey:(NSString *)key {
    [self.defaults setValue:value forKey:[self prefixKey:key]];
}

- (void)removeObjectForKey:(NSString *)key {
    [self.defaults removeObjectForKey:[self prefixKey:key]];
}

- (id)objectForKey:(NSString *)key {
    return [self.defaults objectForKey:[self prefixKey:key]];
}

- (NSString *)stringForKey:(NSString *)key {
    return [self.defaults stringForKey:[self prefixKey:key]];
}

- (NSArray *)arrayForKey:(NSString *)key {
    return [self.defaults arrayForKey:[self prefixKey:key]];
}

- (NSDictionary *)dictionaryForKey:(NSString *)key {
    return [self.defaults dictionaryForKey:[self prefixKey:key]];
}

- (NSData *)dataForKey:(NSString *)key {
    return [self.defaults dataForKey:[self prefixKey:key]];
}

- (NSArray *)stringArrayForKey:(NSString *)key {
    return [self.defaults stringArrayForKey:[self prefixKey:key]];
}

- (NSInteger)integerForKey:(NSString *)key {
    return [self.defaults integerForKey:[self prefixKey:key]];
}

- (float)floatForKey:(NSString *)key {
    return [self.defaults floatForKey:[self prefixKey:key]];
}

- (double)doubleForKey:(NSString *)key {
    return [self.defaults doubleForKey:[self prefixKey:key]];
}

- (BOOL)boolForKey:(NSString *)key {
    return [self.defaults boolForKey:[self prefixKey:key]];
}

- (NSURL *)URLForKey:(NSString *)key {
    return [self.defaults URLForKey:[self prefixKey:key]];
}

- (void)setInteger:(NSInteger)value forKey:(NSString *)key {
    [self.defaults setInteger:value forKey:[self prefixKey:key]];
}

- (void)setFloat:(float)value forKey:(NSString *)key {
    [self.defaults setFloat:value forKey:[self prefixKey:key]];
}

- (void)setDouble:(double)value forKey:(NSString *)key {
    [self.defaults setDouble:value forKey:[self prefixKey:key]];
}

- (void)setBool:(BOOL)value forKey:(NSString *)key {
    [self.defaults setBool:value forKey:[self prefixKey:key]];
}

- (void)setURL:(NSURL *)value forKey:(NSString *)key {
    [self.defaults setURL:value forKey:[self prefixKey:key]];
}

- (void)setObject:(id)value forKey:(NSString *)key {
    [self.defaults setObject:value forKey:[self prefixKey:key]];
}

- (void)migrateUnprefixedKeys:(NSArray *)keys {
    
    for (NSString *key in keys) {
        id value = [self.defaults objectForKey:key];
        if (value) {
            [self.defaults setValue:value forKey:[self prefixKey:key]];
            [self.defaults removeObjectForKey:key];
        }
    }
}

- (void)removeAll {
    for (NSString *key in [[self.defaults dictionaryRepresentation] allKeys]) {
        if ([key hasPrefix:self.keyPrefix]) {
            [self.defaults removeObjectForKey:key];
        }
    }
}

@end

File: Airship/AirshipCore/Source/UAPreferenceDataStore.swift
/* Copyright Airship and Contributors */

/**
 * Preference data store.
 * @note For internal use only. :nodoc:
 */
@objc
public class UAPreferenceDataStore : NSObject {
    private let defaults: UserDefaults
    private let keyPrefix: String

    @objc
    public init(keyPrefix: String) {
        self.defaults = UAPreferenceDataStore.createDefaults(keyPrefix: keyPrefix)
        self.keyPrefix = keyPrefix
        super.init()
    }
    
    class func createDefaults(keyPrefix: String) -> UserDefaults {
        let suiteName = "\(Bundle.main.bundleIdentifier ?? "").airship.settings"
        guard let defaults = UserDefaults(suiteName: suiteName) else {
            AirshipLogger.error("Failed to create defaults \(suiteName)")
            return UserDefaults.standard
        }
        
        for (key, value) in UserDefaults.standard.dictionaryRepresentation() {
            if key.hasPrefix(keyPrefix) {
                defaults.setValue(value, forKey: key)
                UserDefaults.standard.removeObject(forKey: key)
            }
        }
        
        return defaults
    }
    
    func prefixKey(_ key: String) -> String {
        return (keyPrefix) + key
    }

    @objc
    public override func value(forKey key: String) -> Any? {
        return defaults.value(forKey: prefixKey(key))
    }

    @objc
    public override func setValue(_ value: Any?, forKey key: String) {
        defaults.setValue(value, forKey: prefixKey(key))
    }

    @objc
    public func removeObject(forKey key: String) {
        defaults.removeObject(forKey: prefixKey(key))
    }

    @objc
    public func keyExists(_ key: String) -> Bool {
        return object(forKey: key) != nil
    }

    @objc
    public func object(forKey key: String) -> Any? {
        return defaults.object(forKey: prefixKey(key))
    }

    @objc
    public func string(forKey key: String) -> String? {
        return defaults.string(forKey: prefixKey(key))
    }

    @objc
    public func array(forKey key: String) -> [AnyHashable]? {
        return defaults.array(forKey: prefixKey(key)) as? [AnyHashable]
    }

    @objc
    public func dictionary(forKey key: String) -> [AnyHashable : Any]? {
        return defaults.dictionary(forKey: prefixKey(key))
    }

    @objc
    public func data(forKey key: String) -> Data? {
        return defaults.data(forKey: prefixKey(key))
    }

    @objc
    public func stringArray(forKey key: String) -> [AnyHashable]? {
        return defaults.stringArray(forKey: prefixKey(key))
    }

    @objc
    public func integer(forKey key: String) -> Int {
        return defaults.integer(forKey: prefixKey(key))
    }

    @objc
    public func float(forKey key: String) -> Float {
        return defaults.float(forKey: prefixKey(key))
    }

    @objc
    public func double(forKey key: String) -> Double {
        return defaults.double(forKey: prefixKey(key))
    }

    @objc
    public func double(forKey key: String, defaultValue: Double) -> Double {
        if (keyExists(key)) {
            return double(forKey: key)
        } else {
            return defaultValue
        }
    }

    @objc
    public func bool(forKey key: String) -> Bool {
        return defaults.bool(forKey: prefixKey(key))
    }

    @objc
    public func bool(forKey key: String, defaultValue: Bool) -> Bool {
        if keyExists(key) {
            return bool(forKey: key)
        } else {
            return defaultValue
        }
    }

    @objc
    public func url(forKey key: String) -> URL? {
        return defaults.url(forKey: prefixKey(key))
    }

    @objc
    public func setInteger(_ int: Int, forKey key: String) {
        defaults.set(int, forKey: prefixKey(key))
    }

    @objc
    public func setFloat(_ float: Float, forKey key: String) {
        defaults.set(float, forKey: prefixKey(key))
    }

    @objc
    public func setDouble(_ double: Double, forKey key: String) {
        defaults.set(double, forKey: prefixKey(key))
    }

    @objc
    public func setBool(_ bool: Bool, forKey key: String) {
        defaults.set(bool, forKey: prefixKey(key))
    }

    @objc
    public func setURL(_ url: URL?, forKey key: String) {
        defaults.set(url, forKey: prefixKey(key))
    }

    @objc
    public func setObject(_ object: Any?, forKey key: String) {
        defaults.set(object, forKey: prefixKey(key))
    }

    @objc
    func removeAll() {
        for key in defaults.dictionaryRepresentation().keys {
            if key.hasPrefix(keyPrefix) {
                defaults.removeObject(forKey: key)
            }
        }
    }
}
